TestCases/compliance-level-3/0104-allowed-values/0104-allowed-values.dmn

```diff
@@ -0,0 +1,1092 @@
+                <dmn:literalExpression><dmn:text>"baz"</dmn:text></dmn:literalExpression>
+            </dmn:contextEntry>
+        </dmn:context>
+    </dmn:decision>
+
+
+    <!-- compliant assignment to 'context' variable type with constrained list property type -->
+    <dmn:decision name="context_003" id="_context_003">
+        <dmn:variable name="context_003" typeRef="tFooBarListContext"/>
+        <dmn:literalExpression>
+            <dmn:text>{prop1: ["foo", "foo", "bar"]}</dmn:text>
+        </dmn:literalExpression>
+    </dmn:decision>
+
+    <!-- non-compliant assignment to 'context' variable type with constrained list property type -->
+    <dmn:decision name="context_004" id="_context_004">
+        <dmn:variable name="context_004" typeRef="tFooBarListContext"/>
+        <dmn:literalExpression>
+            <dmn:text>{prop1: ["foo", "foo", "baz"]}</dmn:text>
+        </dmn:literalExpression>
+    </dmn:decision>
+
+    <!-- non-compliant assignment to 'context' variable type with constrained property type
+    The intention here is to show that singleton list conversion/equality does not take place
+    inside a structure -->
+    <dmn:decision name="context_005" id="_context_005">
+        <dmn:variable name="context_005" typeRef="tFooBarContext"/>
+        <dmn:literalExpression>
+            <dmn:text>{prop1: ["foo"]}</dmn:text>
+        </dmn:literalExpression>
+    </dmn:decision>
+
+    <!-- non-compliant assignment to 'context' variable type with constrained property type
+        The intention here is to show that singleton list conversion/equality does not take place
+        inside a structure -->
+    <dmn:decision name="context_006" id="_context_006">
+        <dmn:variable name="context_006" typeRef="tFooBarListContext"/>
+        <dmn:literalExpression>
+            <dmn:text>{prop1: "foo"}</dmn:text>
+        </dmn:literalExpression>
+    </dmn:decision>
+
+    <!-- decision echo value for an input that is a context with a constrained property type -->
+    <dmn:decision name="context_input_decision_001" id="_context_input_decision_001">
+        <dmn:variable name="context_input_decision_001"/>
+        <dmn:informationRequirement>
+            <dmn:requiredInput href="#_tFooBarContextInput"/>
+        </dmn:informationRequirement>
+        <dmn:literalExpression>
+            <dmn:text>tFooBarContextInput</dmn:text>
+        </dmn:literalExpression>
+    </dmn:decision>
+
+    <!-- decision with typed literal expression echoing an (untyped) input -->
+    <dmn:decision name="decision_literal_expression_001" id="_decision_literal_expression_001">
+        <dmn:variable name="decision_literal_expression_001" />
+        <dmn:informationRequirement>
+            <dmn:requiredInput href="#_tUntypedInput"/>
+        </dmn:informationRequirement>
+        <dmn:literalExpression typeRef="tFooBarString">
+            <dmn:text>tUntypedInput</dmn:text>
+        </dmn:literalExpression>
```

Comment:
TCK 2021-07-15 
wfrom the DMN1.3 spec

> If an instance of Expression that defines the output of a Decision element includes a typeRef, the referenced type SHALL be the same as the type of the containing Decision element.

So this DMN is not conformant in the way it is right now.

We will check if we have some general provision in the spec which mandates that for any Expression defining a typeRef, the typeConformance shall be expressed.
Specific example.
For a ContextEntry to specify a typeRef, that the entry value must conform to the typeRef.

Before going forward with this kind of test would be meaningful to clarify the standing ground for type conformance check in the Expression.

We will also check, and eventually raise to the RTF group, if we will see that the DMN specification is missing an "analogous" and corresponding statement for the Expression which are the body of a FunctionDefinition encapsulatedLogic of BKM, where we already agreed is the return type of the function definition.

Reply:
HI @tarilabs - re the quote above.  That sure feels out of date.  If anything, they should _conform_ I guess.  But, as it is, I think you are saying that if a decision expression has a typeref, then the containing decision must also have that exact typeRef also.  Doesn't seems right (IMO).

Reply:
I am still under the impression this DMN model is _not_ conformant based on https://github.com/dmn-tck/tck/pull/422#discussion_r670424829 and our mutual understanding of BKM and its contained typeRefs (in the variable and the body).

> If an instance of Expression that defines the output of a Decision element includes a typeRef, the referenced type SHALL be the same as the type of the containing Decision element.

The spec as it stands right now is quite clear: if a Decision has an expression which defines a typeRef (eg: the Decision's LiteralExpression, the Decision's boxed context, etc) then the same typeRef SHALL be the same as the type of the Decision (in the Decision's variable, that is).

On this premise, one could say for Decision you could just specify the typeRef in the Decision's variable (as specifying the same typeRef in the Decision's boxed expression is redundant).

I am open to revise this perspective, but it would require a change in the spec, for which I'm not sure the rationale... when would you want to specify a _different_ typeRef for the Decision's variable Vs the Decision's boxed expression?